test/metabase/actions/test_util.clj

```diff
@@ -117,83 +117,89 @@
+                               options-map))]
+        {:action-id action-id :model-id model-id})
+    :implicit
+    (let [action-id (action/insert! (merge
+                                        {:type :implicit
+                                         :name "Update Example"
+                                         :kind "row/update"
+                                         :model_id model-id}
+                                        options-map))]
+        {:action-id action-id :model-id model-id})
+
+    :http
+    (mt/with-model-cleanup [Action]
+      (let [action-id (action/insert! (merge
+                                        {:type :http
+                                         :name "Echo Example"
+                                         :template {:url (client/build-url "testing/echo[[?fail={{fail}}]]" {})
+                                                    :method "POST"
+                                                    :body "{\"the_parameter\": {{id}}}"
+                                                    :headers "{\"x-test\": \"{{id}}\"}"}
+                                         :parameters [{:id "id"
+                                                       :type "number"
+                                                       :target [:template-tag "id"]}
+                                                      {:id "fail"
+                                                       :type "text"
+                                                       :target [:template-tag "fail"]}]
+                                         :response_handle ".body"
+                                         :model_id model-id}
+                                        options-map))]
+        {:action-id action-id :model-id model-id}))))
 
 (defmacro with-action
   "Execute `body` with a newly created Action.
    `binding-form` is a returned map with key `:action-id`, and `:query-action-card-id` for QueryActions.
    `options-map` contains overrides for the action. Defaults to a sane QueryAction.
 
-    (with-action [{:keys [action-id], :as context} {:type :http :name \"Temp HTTP Action\"}]
-      (do-something))"
+   (with-action [{:keys [action-id], :as context} {:type :http :name \"Temp HTTP Action\"}]
+   (do-something))"
   {:style/indent 1}
-  [[binding-form options-map] & body]
-  `(do-with-action
-     (merge {:type :query} ~options-map)
-     (fn [~binding-form] ~@body)))
+  [binding-forms-and-option-maps & body]
+  (let [model-id (gensym "model-id_")]
+    `(do
+       (initialize/initialize-if-needed! :web-server)
+       (mt/with-temp* [Card [{~model-id :id} {:dataset true :dataset_query (mt/mbql-query categories)}]]
+         (mt/with-model-cleanup [Action]
+           (let [~@(mapcat (fn [[binding-form option-map]]
+                             [binding-form `(do-with-action (merge {:type :query} ~option-map) ~model-id)])
+                           (partition-all 2 binding-forms-and-option-maps))]
```

Comment:
Would it make sense to insist on an even number of elements in `binding-forms-and-option-maps`? Only the last one can be omitted, otherwise you get a mess. Would you encourage `(with-action [{id :action-id}] ...)`?

Reply:
I think it makes sense for pairs otherwise I don't think it should be called a binding form. I would discourage that example as it feels very non-idiomatic. `(let [x nil] ...) vs (let [x] ...)`